File: Simulator.m
classdef Simulator < handle
    properties(Constant)
        gamma = 267522187.44 % rad/s/T
        gamma_bar = 42577478.518 % Hz/T
    end
    
    properties
        M0
        T1
        T2
        positions
        delta_omega
    end
    
    properties (SetAccess = private)
        magnetization_
    end

    properties (Dependent)
        magnetization
    end
    
    methods
        function self = Simulator(M0, T1, T2, positions, delta_omega)
            self.M0 = M0;
            self.T1 = T1;
            self.T2 = T2;
            self.positions = positions;
            
            if (~exist('delta_omega', 'var'))
                delta_omega = zeros(size(M0));
            end
            self.delta_omega = delta_omega;

            self.magnetization_ = zeros([4, 1, length(M0)]);
            self.magnetization_(3, :, :) = M0;
            self.magnetization_(4, :, :) = 1;
        end
        
        function M = get.magnetization(self)
            M = self.magnetization_(1:3, :, :) ./ self.magnetization_(4:4, :, :);
            M = squeeze(M);
        end

        function pulse(self, angle, phase)
            if length(angle) == 1
                angle = angle * ones(size(self.M0));
            end
            
            if nargin <= 2
                phase = zeros(size(angle));
            end

            operator = zeros(4, 4, length(self.M0));
            for i = 1:length(self.M0)
                operator(1:3, 1:3, i) = compute_rotation_matrix( ...
                    [cos(phase(i)), sin(phase(i)), 0], angle(i));
                operator(4, 4, i) = 1;
            end
            
            self.magnetization_ = pagemtimes(operator, self.magnetization_);
        end

        function free_precession(self, duration)
            angle = 2*pi*duration*self.delta_omega;
            operator = zeros(4, 4, length(self.M0));
            for i = 1:length(self.M0)
                operator(1:3, 1:3, i) = compute_rotation_matrix( ...
                    [0, 0, 1], angle(i));
                operator(4, 4, i) = 1;
            end

            self.magnetization_ = pagemtimes(operator, self.magnetization_);
        end

        function relaxation(self, duration)
            E1 = exp(-duration ./ self.T1);
            E2 = exp(-duration ./ self.T2);

            operator = zeros(4, 4, length(self.M0));
            operator(1, 1, :) = E2;
            operator(2, 2, :) = E2;
            operator(3, 3, :) = E1;
            operator(4, 4, :) = 1;
            operator(3, 4, :) = self.M0 .* (1-E1);

            self.magnetization_ = pagemtimes(operator, self.magnetization_);
        end

        function idle(self, duration)
            self.relaxation(duration);
            self.free_precession(duration);
        end

        function gradient(self, amplitude, duration)
            self.relaxation(duration);
            
            if size(amplitude, 1) == 1
                if size(self.positions, 1) ~= 1
                    amplitude = repmat(amplitude, size(self.positions, 1), 1);
                else
                    amplitude = repmat(amplitude, 1, size(self.positions, 2));
                end
            end
            
            if size(self.positions, 1) ~= 1
                delta_B = dot(self.positions, amplitude, 2);
            else
                delta_B = dot(self.positions, amplitude, 1);
            end
            
            delta_omega = self.delta_omega + self.gamma_bar * delta_B;
            angle = 2*pi * duration * delta_omega;
            
            operator = zeros(4, 4, length(self.M0));
            for i = 1:length(self.M0)
                operator(1:3, 1:3, i) = compute_rotation_matrix( ...
                    [0, 0, 1], angle(i));
                operator(4, 4, i) = 1;
            end
            
            self.magnetization_ = pagemtimes(operator, self.magnetization_);
        end
    end
end

function R = compute_rotation_matrix(k, theta)
    K = [
        [0, -k(3), k(2)]
        [k(3), 0, -k(1)]
        [-k(2), k(1), 0]];
    R = eye(3) + K * sin(theta) + K * K * (1-cos(theta));
end

File: source/simulator.py
import matplotlib.pyplot
import numpy

class Simulator(object):
    gamma = 267522187.44 # rad/s/T
    gamma_bar = 42577478.518 # Hz/T
        
    def __init__(self, M0, T1, T2, positions, delta_omega=None):
        M0, T1, T2, positions = [
            numpy.asarray(x) for x in [M0, T1, T2, positions]]
        
        if delta_omega is None:
            delta_omega = numpy.zeros(M0.shape)
        else:
            delta_omega = numpy.asarray(delta_omega)
        
        if any(x.ndim != 1 for x in [T1, T2, delta_omega]):
            raise Exception("M0, T1, T2, and delta_omega must be 1D")
        if any(M0.shape != x.shape for x in [T1, T2, delta_omega]):
            raise Exception("All inputs must be have the same size")
        if M0.shape != positions.shape[:1]:
            raise Exception("All inputs must be have the same size")
        
        self.M0 = numpy.array(M0)
        self._magnetization = numpy.zeros((len(M0), 4))
        self._magnetization[:, 2] = M0
        self._magnetization[:, 3] = 1
        
        self.T1 = numpy.array(T1)
        self.T2 = numpy.array(T2)
        self.positions = numpy.array(positions)
        self.delta_omega = delta_omega
    
    @property
    def magnetization(self):
        # NOTE: self.M[:, 3] will always be one
        return self._magnetization[:,:3] / self._magnetization[:,3:]
    
    def pulse(self, angle, phase=None):
        angle = numpy.atleast_1d(angle)
        phase = (
            numpy.atleast_1d(phase)
            if phase is not None else numpy.zeros_like(angle))
        operator = numpy.zeros((*angle.shape, 4, 4))
        operator[:, :3, :3] = numpy.array([
            compute_rotation_matrix([numpy.cos(p), numpy.sin(p), 0], a)
            for a, p in zip(angle, phase)])
        operator[:, 3, 3] = 1
        
        self._magnetization = numpy.einsum(
            "nij,nj->ni", operator, self._magnetization)
    
    def free_precession(self, duration):
        operator = precession(self.delta_omega, duration)
        self._magnetization = numpy.einsum(
            "nij,nj->ni", operator, self._magnetization)
    
    def relaxation(self, duration):
        operator = relaxation(self.M0, self.T1, self.T2, duration)
        self._magnetization = numpy.einsum(
            "nij,nj->ni", operator, self._magnetization)
    
    def idle(self, duration):
        op1 = relaxation(self.M0, self.T1, self.T2, duration)
        op2 = precession(self.delta_omega, duration)
        operator = op2 @ op1
        self._magnetization = numpy.einsum(
            "nij,nj->ni", operator, self._magnetization)
    
    def gradient(self, amplitude, duration):
        op1 = relaxation(self.M0, self.T1, self.T2, duration)
        
        delta_B = numpy.dot(self.positions, numpy.array(amplitude))
        delta_omega = self.gamma_bar * delta_B
        op2 = precession(delta_omega+self.delta_omega, duration)
        
        operator = op2 @ op1
        self._magnetization = numpy.einsum(
            "nij,nj->ni", operator, self._magnetization)

def precession(delta_omega, duration):
    angle = 2*numpy.pi*duration*delta_omega
    
    operator = numpy.zeros((*angle.shape, 4, 4))
    operator[:, :3, :3] = numpy.array(
        [compute_rotation_matrix([0, 0, 1], x) for x in angle])
    operator[:, 3, 3] = 1
    
    return operator

def relaxation(M0, T1, T2, duration):
    E1 = numpy.exp(-duration/T1)
    E2 = numpy.exp(-duration/T2)
    
    operator = numpy.zeros((len(T1), 4, 4))
    operator[:, 0, 0] = E2
    operator[:, 1, 1] = E2
    operator[:, 2, 2] = E1
    operator[:, 3, 3] = 1
    operator[:, 2, 3] = M0*(1-E1)
    
    return operator

def compute_rotation_matrix(k, theta):
    """ Return the matrix corresponding to the rotation around
        axis k by an angle theta (in radians).
    """
    
    K = numpy.array([
        [0, -k[2], k[1]],
        [k[2], 0, -k[0]],
        [-k[1], k[0], 0],
    ])
    R = (
        numpy.identity(3)
        + K*numpy.sin(theta)
        + numpy.linalg.matrix_power(K, 2) * (1-numpy.cos(theta)))
    return R
